nomad/structs/csi.go

```diff
@@ -652,17 +652,50 @@ func (p *CSIPlugin) AddPlugin(nodeID string, info *CSIInfo) {
 // DeleteNode removes all plugins from the node. Called from state.DeleteNode in a
 // transaction
 func (p *CSIPlugin) DeleteNode(nodeID string) {
+	p.DeleteNodeType(nodeID, CSIPluginTypeMonolith)
+}
+
+// DeleteNodeType deletes all plugins of type from the node. Called from deleteJobFromPlugin
+// during job Deregistration
+func (p *CSIPlugin) DeleteNodeType(nodeID string, pluginType CSIPluginType) {
+	switch pluginType {
+	case CSIPluginTypeController:
+		prev, ok := p.Controllers[nodeID]
+		if ok && prev.Healthy {
+			p.ControllersHealthy -= 1
+		}
+		delete(p.Controllers, nodeID)
+
+	case CSIPluginTypeNode:
+		prev, ok := p.Nodes[nodeID]
+		if ok && prev.Healthy {
+			p.NodesHealthy -= 1
+		}
+		delete(p.Nodes, nodeID)
+
+	case CSIPluginTypeMonolith:
+		p.DeleteNodeType(nodeID, CSIPluginTypeController)
+		p.DeleteNodeType(nodeID, CSIPluginTypeNode)
+	}
+}
+
+// DeleteAlloc removes the fingerprint info for the allocation
+func (p *CSIPlugin) DeleteAlloc(allocID, nodeID string) {
 	prev, ok := p.Controllers[nodeID]
-	if ok && prev.Healthy {
-		p.ControllersHealthy -= 1
+	if ok && prev.AllocID == allocID {
```

Comment:
Just thinking about this because I know we store `nil`s in these fields... should we nil-check `prev` or are we guaranteed it'll always be non-nil in this call?

Reply:
Guaranteed is interesting... it is always non-nil now because of the call path, but there's not a great way to enforce that. Maybe it should return an error in the case it gets a nil

Reply:
done